runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/retries/impl/StandardRetryPolicy.kt

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+package aws.smithy.kotlin.runtime.retries.impl
+
+import aws.smithy.kotlin.runtime.ClientException
+import aws.smithy.kotlin.runtime.SdkBaseException
+import aws.smithy.kotlin.runtime.ServiceException
+import aws.smithy.kotlin.runtime.ServiceException.*
+import aws.smithy.kotlin.runtime.retries.RetryDirective
+import aws.smithy.kotlin.runtime.retries.RetryErrorType
+import aws.smithy.kotlin.runtime.retries.RetryPolicy
+
+/**
+ * A standard retry policy which attempts to derive information from the Smithy exception hierarchy.
+ */
+open class StandardRetryPolicy : RetryPolicy<Any?> {
+    /**
+     * Evaluate the given retry attempt.
+     */
+    override fun evaluate(result: Result<Any?>): RetryDirective = when {
+        result.isSuccess -> RetryDirective.TerminateAndSucceed
+        else -> evaluate(result.exceptionOrNull()!!)
+    }
+
+    private fun evaluate(ex: Throwable): RetryDirective =
+        (ex as? SdkBaseException)?.run(::evaluateBaseException)
```

Comment:
comment

I find this to be a mess to read and debug. Isn't there a bit cleaner (if not more verbose) way of expressing this?

I believe you had this as a `when` expression at one point in time that IMO is much easier to read and set breakpoints on.

Reply:
`when` would be preferable if it allowed fall-through. The key here is that we want to try each strategy until we find something that works.

That said, yes there are alternate ways of writing this which may be cleaner...like an actual list of strategies to try in a loop.